### PR TITLE
Auto restarted Jupyter kernel

### PR DESCRIPTION
**Pull request by SmartManoj**
Solves #1616, #1627

1. Auto restarted Jupyter kernel. Sometimes LLM is ignoring that.  
https://github.com/OpenDevin/OpenDevin/issues/1616#issuecomment-2112150465

2. Context length is increased quickly due to the output of the package installation. So, changed the output of pip installs to a single line.
https://github.com/OpenDevin/OpenDevin/issues/1616#issuecomment-2106134659

3. Changed regex to non-greedy versions for the following LLM output
```
Install the module using
<execute_ipython>
%pip install matplotlib
</execute_ipython>

Then import using this ...
<execute_ipython>
import matplotlib.pyplot as plt
</execute_ipython>
```

4. Refactored `execute_cli`

image

### CODE REVIEW

**Review by yufansong — APPROVED** [review]
LGTM

**Review by rbren — CHANGES_REQUESTED** [review]
Blocking so that @xingyaoww has time to look at the prompt changes.

Ping me in Slack when it's ready and I'll unblock

File: agenthub/codeact_agent/prompt.py
Comment (SmartManoj) [inline agenthub/codeact_agent/prompt.py:41]: @xingyaoww The assistant should stop <execute>. 
How it will stop?
--
provide an answer when they have already obtained the answer from the execution result.
It's not correct when the code provides dynamic output. Also, the LLM is aware of that. Do you have any examples for the repetition?

**Review by xingyaoww — CHANGES_REQUESTED** [review]
Mostly LGTM! Can you bump the CodeActAgent's version since we are modifying prompt?

**Review by rbren — APPROVED** [review]

File: agenthub/codeact_agent/prompt.py
Comment (xingyaoww) [reply agenthub/codeact_agent/prompt.py:41]: > How it will stop?

Most likely the assistant will just output pure natural language text that won't get parsed as action?

**Review by xingyaoww — APPROVED** [review]
LGTM

**Comment by neubig** [comment]
@SmartManoj : this is ready to merge other than codecov. Could you add tests to the modified parts?

**Comment by SmartManoj** [comment]
poetry commands fail with [Errno 2] No such file or directory
Similar issue: https://github.com/python-poetry/poetry/issues/6841
But without running poetry also causes this issue.
https://bugs.launchpad.net/apport/+bug/1977954
image
Patched apport. Still, issue is there.
image

**Comment by enyst** [comment]
I guess it suddenly doesn't find the right python.

You may want to try the suggestions in the comments to the poetry issue you linked above, to fix/clean the poetry virtual environments. Just not the one comment with rm -rf [poetry-config-variable] because that one seems to be risky! 😅

**Comment by SmartManoj** [comment]
Seems poetry is not the problem. Because in the last image error occurs in `os.getcwd()` ; 

--

In the last test, prompts are equal.
https://www.diffchecker.com/RbS3al9o/

**Comment by enyst** [comment]
The diffchecker link seems to show a difference with the separator. @li-boxuan may have already seen something like that happening? 

I updated the branch to make sure it gets all fixes, I believe there have been changes to integration tests very recently.

**Comment by li-boxuan** [comment]
Do you mind reviewing https://github.com/OpenDevin/OpenDevin/pull/1855 and merging it if you approve it?

That PR would solve the separator problem.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by li-boxuan** [comment]
Note: after merging main branch, you need to revert/remove previous regeneration commits and rerun the regenerate script, otherwise nothing happens if the test is passing.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

File: tests/integration/regenerate.sh
Comment (li-boxuan) [inline tests/integration/regenerate.sh:27]:
```suggestion
  "Install and import pymsgbox==1.0.9 and print it's version in /workspace/test.txt."
```

The latter part is already included in `regenerate_with_llm` function.

Comment (SmartManoj) [reply tests/integration/regenerate.sh]: Oops. Forget to commit that change and compared with local mock. That's why the diff shows no character changes.

**Pull request merged**
